File: Gradely/Support/SchoolDirectorySearch.swift
import Foundation

enum SchoolDirectorySearch {
    static func results(
        for query: String,
        in schools: [SchoolDirectorySchool],
        limit: Int = 8
    ) -> [SchoolDirectorySchool] {
        let trimmedQuery = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedQuery.isEmpty else { return [] }

        let normalizedQuery = normalized(trimmedQuery)
        let tokens = normalizedQuery
            .split(whereSeparator: \.isWhitespace)
            .map(String.init)

        return schools
            .compactMap { school -> (school: SchoolDirectorySchool, score: Int)? in
                guard let score = score(school, query: normalizedQuery, tokens: tokens) else {
                    return nil
                }
                return (school, score)
            }
            .sorted { lhs, rhs in
                if lhs.score != rhs.score { return lhs.score < rhs.score }
                if lhs.school.trimmedTown.localizedCompare(rhs.school.trimmedTown) != .orderedSame {
                    return lhs.school.trimmedTown.localizedCompare(rhs.school.trimmedTown) == .orderedAscending
                }
                return lhs.school.trimmedName.localizedCompare(rhs.school.trimmedName) == .orderedAscending
            }
            .prefix(limit)
            .map(\.school)
    }

    static func normalized(_ value: String) -> String {
        value
            .folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Locale(identifier: "cs_CZ"))
            .lowercased()
    }

    private static func acronym(for value: String) -> String {
        normalized(value)
            .split { !$0.isLetter && !$0.isNumber }
            .compactMap(\.first)
            .map(String.init)
            .joined()
    }

    private static func score(
        _ school: SchoolDirectorySchool,
        query: String,
        tokens: [String]
    ) -> Int? {
        let name = normalized(school.name)
        let town = normalized(school.town)
        let url = normalized(school.schoolURL)
        let nameAcronym = acronym(for: name)
        let searchable = [name, town, url, nameAcronym].joined(separator: " ")

        guard tokens.allSatisfy({ searchable.contains($0) }) else { return nil }

        if name == query { return 0 }
        if nameAcronym == query || nameAcronym.hasPrefix(query) { return 5 }
        if name.hasPrefix(query) { return 10 }
        if town.hasPrefix(query) { return 20 }
        if name.contains(query) { return 30 }
        if town.contains(query) { return 40 }
        if url.contains(query) { return 50 }
        return 60
    }
}

File: android/core-domain/src/main/kotlin/com/bukovinafilip/gradey/domain/SchoolDirectorySearch.kt
package com.bukovinafilip.gradey.domain

import com.bukovinafilip.gradey.model.SchoolDirectorySchool
import java.text.Collator
import java.text.Normalizer
import java.util.Locale

object SchoolDirectorySearch {
    fun results(
        query: String,
        schools: List<SchoolDirectorySchool>,
        limit: Int = 8,
    ): List<SchoolDirectorySchool> {
        val normalizedQuery = normalize(query.trim())
        if (normalizedQuery.isEmpty() || limit <= 0) return emptyList()
        val tokens = normalizedQuery.split(Regex("\\s+")).filter(String::isNotEmpty)
        val collator = Collator.getInstance(CZECH_LOCALE)

        return schools
            .mapNotNull { school -> score(school, normalizedQuery, tokens)?.let { school to it } }
            .sortedWith { left, right ->
                when {
                    left.second != right.second -> left.second.compareTo(right.second)
                    collator.compare(left.first.trimmedTown, right.first.trimmedTown) != 0 ->
                        collator.compare(left.first.trimmedTown, right.first.trimmedTown)
                    else -> collator.compare(left.first.trimmedName, right.first.trimmedName)
                }
            }
            .take(limit)
            .map { it.first }
    }

    fun normalize(value: String): String = Normalizer
        .normalize(value.lowercase(CZECH_LOCALE), Normalizer.Form.NFD)
        .replace(COMBINING_MARKS, "")

    private fun acronym(value: String): String = normalize(value)
        .split(NON_ALPHANUMERIC)
        .filter(String::isNotEmpty)
        .mapNotNull(String::firstOrNull)
        .joinToString(separator = "")

    private fun score(
        school: SchoolDirectorySchool,
        query: String,
        tokens: List<String>,
    ): Int? {
        val name = normalize(school.name)
        val town = normalize(school.town)
        val url = normalize(school.schoolURL)
        val nameAcronym = acronym(name)
        val searchable = "$name $town $url $nameAcronym"
        if (!tokens.all(searchable::contains)) return null

        return when {
            name == query -> 0
            nameAcronym == query || nameAcronym.startsWith(query) -> 5
            name.startsWith(query) -> 10
            town.startsWith(query) -> 20
            name.contains(query) -> 30
            town.contains(query) -> 40
            url.contains(query) -> 50
            else -> 60
        }
    }

    private val CZECH_LOCALE = Locale.forLanguageTag("cs-CZ")
    private val COMBINING_MARKS = Regex("\\p{M}+")
    private val NON_ALPHANUMERIC = Regex("[^\\p{L}\\p{N}]+")
}
